services/src/main/java/org/apache/unomi/services/services/ProfileServiceImpl.java

```diff
@@ -39,9 +39,99 @@
 import java.io.IOException;
 import java.net.URL;
 import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class ProfileServiceImpl implements ProfileService, SynchronousBundleListener {
 
+    private static class PropertyTypes {
```

Comment:
Maybe we could add a comment on this class that explains why we are using copies on write: basically to make sure that we don't impact performance with contention and that we insure a consistent state as much as possible.

Reply:
OK, I will do that.